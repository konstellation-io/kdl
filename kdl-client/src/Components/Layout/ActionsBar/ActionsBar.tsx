import React, { FC } from 'react';

import { HorizontalBar } from 'kwc';
import cx from 'classnames';
import styles from './ActionsBar.module.scss';

type Props = {
  centerActions?: boolean;
  children: JSX.Element | JSX.Element[];
};
const ActionsBar: FC<Props> = ({ children, centerActions = false }) => {
  return (
    <div className={styles.container}>
      <HorizontalBar
        className={cx(styles.actions, { [styles.center]: centerActions })}
      >
        {children}
      </HorizontalBar>
    </div>
  );
};

export default ActionsBar;
