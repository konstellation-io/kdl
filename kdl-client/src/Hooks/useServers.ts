import { StoreKey, StoreUpdate } from 'types/store';
import { useEffect, useState } from 'react';

import { Action } from 'Pages/Servers/components/Server/Server';
import IconStart from '@material-ui/icons/PlayArrow';
import IconStop from '@material-ui/icons/Stop';
import { LocalServerStates } from 'Pages/Servers/components/Server/LocalServer';
import { RemoteServerStates } from 'Pages/Servers/components/Server/RemoteServer';
import { ipcRenderer } from 'electron';

export enum ServerType {
  LOCAL = 'local',
  REMOTE = 'remote',
}
export type Server = {
  id: string;
  name: string;
  type: ServerType;
  state: LocalServerStates | RemoteServerStates;
  url?: string;
};

export type AddServer = {
  name: string;
  type: ServerType;
  url?: string;
};

function useServers() {
  const [servers, setServers] = useState<Server[]>([]);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    function onStoreUpdate(_: unknown, { key, value }: StoreUpdate) {
      if (key === StoreKey.SERVERS) {
        setServers(value);
      }
    }

    ipcRenderer.send('subscribeToValue', StoreKey.SERVERS);
    ipcRenderer.on('subscribeToValueReply', onStoreUpdate);

    async function fetchData() {
      const storeServers = await ipcRenderer.invoke(
        'getStoreValue',
        StoreKey.SERVERS
      );
      setServers(storeServers);
      setLoading(false);
    }
    fetchData();

    return () => {
      ipcRenderer.removeListener('subscribeToValueReply', onStoreUpdate);
    };
  }, []);

  function getServer(id: string): Server | undefined {
    return servers.find((server) => server.id === id);
  }

  function getServerActions(
    state: LocalServerStates | RemoteServerStates,
    id: string
  ) {
    let actions: Action[] = [];

    switch (state) {
      case LocalServerStates.STARTED:
        actions = [
          {
            label: 'STOP',
            Icon: IconStop,
            onClick: () => ipcRenderer.send('stopLocalServer', id),
          },
        ];
        break;
      case LocalServerStates.STOPPED:
        actions = [
          {
            label: 'START',
            Icon: IconStart,
            onClick: () => ipcRenderer.send('startLocalServer', id),
          },
        ];
        break;
      default:
        break;
    }

    return actions;
  }

  return {
    servers,
    loading,
    getServer,
    getServerActions,
  };
}

export default useServers;
