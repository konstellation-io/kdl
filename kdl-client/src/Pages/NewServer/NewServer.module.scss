@import 'Styles/variables';

.container {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-between;
  width: 100%;
  height: 100%;

  .actionBar {
    width: 100%;
  }

  .content {
    margin-top: 8 * $grid-unit;
    display: flex;
    justify-content: center;
    align-items: center;
    flex-direction: column;

    .subtitle {
      margin-top: 2 * $grid-unit;
      width: 38 * $grid-unit;
      text-align: center;
    }

    .servers {
      display: flex;
      justify-content: space-between;
      pointer-events: none;
      height: 100%;
      width: 100%;

      .server {
        pointer-events: auto;
      }
    }
  }
}
