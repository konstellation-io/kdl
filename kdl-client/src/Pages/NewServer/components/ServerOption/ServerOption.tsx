import { Button } from 'kwc';
import { Link } from 'react-router-dom';
import React from 'react';
import containerStyles from '../../NewServer.module.scss';
import cx from 'classnames';
import serverStyles from '../Server/Server.module.scss';
import styles from './ServerOption.module.scss';

type Props = {
  title: string;
  subtitle: string;
  actionLabel: string;
  to: string;
  Server: JSX.Element;
};

function ServerOption({ title, subtitle, actionLabel, to, Server }: Props) {
  return (
    <Link to={to}>
      <div
        className={cx(
          styles.container,
          serverStyles.hoverContainer,
          containerStyles.server
        )}
      >
        <div>{Server}</div>
        <p className={styles.title}>{title}</p>
        <p className={styles.subtitle}>{subtitle}</p>
        <div className={styles.button}>
          <Button label={actionLabel} primary />
        </div>
      </div>
    </Link>
  );
}

export default ServerOption;
