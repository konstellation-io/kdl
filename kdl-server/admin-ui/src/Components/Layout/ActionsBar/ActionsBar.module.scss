@import 'Styles/variables';
@import 'Styles/mixins';
@import 'Styles/colors';
@import 'Styles/shadows';

.container {
  @include shadow(2);

  height: 5 * $grid-unit;
  flex-shrink: 0;
}

.actions {
  background-color: palette(base, 400);
  padding: 0 calc(50% - (#{$column-page-content-width / 2}));
  justify-content: flex-end;

  &.center {
    padding: 0;
    justify-content: center;
  }
}
