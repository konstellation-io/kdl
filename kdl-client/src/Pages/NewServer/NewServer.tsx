import Server, { DIRECTION } from './components/Server/Server';

import ActionsBar from '../../Components/Layout/ActionsBar/ActionsBar';
import { Button } from 'kwc';
import ROUTE from 'Constants/routes';
import React from 'react';
import ServerOption from './components/ServerOption/ServerOption';
import styles from './NewServer.module.scss';
import useServers from 'Hooks/useServers';

function NewServer() {
  const { servers } = useServers();
  const hasServers = !!servers.length;

  return (
    <div className={styles.container}>
      <div className={styles.content}>
        <h1>Add a Server</h1>
        <h3 className={styles.subtitle}>
          Cras quis nulla commodo, aliquam lectus sed, blandit augue. Cras
          ullamcorper bibendum bibendum.
        </h3>
        <div className={styles.servers}>
          <ServerOption
            title="Connect to a Remote Server"
            subtitle="Mauris non tempor quam, et lacinia sapien. Mauris accumsan eros eget libero posuere vulputate. "
            actionLabel="CONNECT"
            to={ROUTE.CONNECT_TO_REMOTE_SERVER}
            Server={<Server direction={DIRECTION.UP} />}
          />
          <ServerOption
            title="Install a Local Server"
            subtitle="Fusce vehicula dolor arcu, sit amet blandit dolor mollis nec. Donec viverra eleifend lacus, vitae."
            actionLabel="INSTALL"
            to={ROUTE.CHECK_LOCAL_SERVER_REQUIREMENTS}
            Server={<Server direction={DIRECTION.DOWN} />}
          />
        </div>
      </div>
      {hasServers && (
        <ActionsBar className={styles.actionBar} centerActions>
          <Button label="CANCEL" to={ROUTE.HOME} />
        </ActionsBar>
      )}
    </div>
  );
}

export default NewServer;
