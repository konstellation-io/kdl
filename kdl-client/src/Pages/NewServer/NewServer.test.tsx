import ActionsBar from '../../Components/Layout/ActionsBar/ActionsBar';
import NewServer from './NewServer';
import React from 'react';
import ServerOption from './components/ServerOption/ServerOption';
import { shallow } from 'enzyme';
import useServers from 'Hooks/useServers';

jest.mock('Hooks/useServers');

let component;

describe('NewServer page without servers', () => {
  beforeEach(() => {
    useServers.mockImplementationOnce(() => ({
      servers: [],
    }));

    component = shallow(<NewServer />);
  });

  test('Component match snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('show right texts', () => {
    expect(component.contains('Add a Server')).toBeTruthy();
  });

  it('has two options', () => {
    expect(component.find(ServerOption).length).toBe(2);
  });

  it('does not show actions when there are no servers', () => {
    expect(component.find(ActionsBar).isEmptyRender()).toBeTruthy();
  });
});

describe('NewServer page with servers', () => {
  beforeEach(() => {
    useServers.mockImplementationOnce(() => ({
      servers: [{ url: 'some.server' }],
    }));

    component = shallow(<NewServer />);
  });

  test('Component match snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('shows actions when there are servers', () => {
    useServers.mockImplementation(() => ({
      servers: [{ url: 'some.server' }],
    }));

    component = shallow(<NewServer />);

    expect(component.find(ActionsBar).isEmptyRender()).toBeFalsy();
  });
});
