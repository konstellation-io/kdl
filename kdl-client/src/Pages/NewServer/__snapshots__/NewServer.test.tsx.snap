// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewServer page with servers Component match snapshot 1`] = `
<div
  className="container"
>
  <div
    className="content"
  >
    <h1>
      Add a Server
    </h1>
    <h3
      className="subtitle"
    >
      Before working on a project, you must first connect to a local or remote server.
    </h3>
    <div
      className="servers"
    >
      <ServerOption
        Server={
          <Server
            direction="up"
          />
        }
        actionLabel="CONNECT"
        subtitle="Remote servers are ready and might contain active projects. You need access privileges to the Server before connecting to it."
        title="Connect to a Remote Server"
        to="/connect-to-remote-server"
      />
      <ServerOption
        Server={
          <Server
            direction="down"
          />
        }
        actionLabel="INSTALL"
        subtitle="You can work locally by installing a KDL Server in your machine. Only one local Server can be installed."
        title="Install a Local Server"
        to="/check-local-server-requirements"
      />
    </div>
  </div>
  <ActionsBar
    centerActions={true}
    className="actionBar"
  >
    <P
      label="CANCEL"
      to="/"
    />
  </ActionsBar>
</div>
`;

exports[`NewServer page without servers Component match snapshot 1`] = `
<div
  className="container"
>
  <div
    className="content"
  >
    <h1>
      Add a Server
    </h1>
    <h3
      className="subtitle"
    >
      Before working on a project, you must first connect to a local or remote server.
    </h3>
    <div
      className="servers"
    >
      <ServerOption
        Server={
          <Server
            direction="up"
          />
        }
        actionLabel="CONNECT"
        subtitle="Remote servers are ready and might contain active projects. You need access privileges to the Server before connecting to it."
        title="Connect to a Remote Server"
        to="/connect-to-remote-server"
      />
      <ServerOption
        Server={
          <Server
            direction="down"
          />
        }
        actionLabel="INSTALL"
        subtitle="You can work locally by installing a KDL Server in your machine. Only one local Server can be installed."
        title="Install a Local Server"
        to="/check-local-server-requirements"
      />
    </div>
  </div>
</div>
`;
